agent/engine/docker_task_engine.go

```diff
@@ -466,14 +472,36 @@ func (engine *DockerTaskEngine) GetTaskByArn(arn string) (*api.Task, bool) {
 }
 
 func (engine *DockerTaskEngine) pullContainer(task *api.Task, container *api.Container) DockerContainerMetadata {
-	log.Info("Pulling container", "task", task, "container", container)
+	if engine.enabledConcurrentPull {
+		return engine.concurrentPull(task, container)
+	} else {
+		return engine.serialPull(task, container)
+	}
+}
+
+func (engine *DockerTaskEngine) concurrentPull(task *api.Task, container *api.Container) DockerContainerMetadata {
+	log.Info("Pulling container concurrently", "task", task, "container", container)
+	seelog.Debugf("Attempting to obtain ImagePullDeleteLock to pull image - %s", container.Image)
+
+	ImagePullDeleteLock.RLock()
+	defer seelog.Debugf("Released ImagePullDeleteLock after pulling image - %s", container.Image)
+	defer ImagePullDeleteLock.RUnlock()
+
+	return engine.pullAndUpdateContainerReference(task, container)
+}
+
+func (engine *DockerTaskEngine) serialPull(task *api.Task, container *api.Container) DockerContainerMetadata {
+	log.Info("Pulling container serially", "task", task, "container", container)
 	seelog.Debugf("Attempting to obtain ImagePullDeleteLock to pull image - %s", container.Image)
 
 	ImagePullDeleteLock.Lock()
-	seelog.Debugf("Obtained ImagePullDeleteLock to pull image - %s", container.Image)
 	defer seelog.Debugf("Released ImagePullDeleteLock after pulling image - %s", container.Image)
 	defer ImagePullDeleteLock.Unlock()
 
+	return engine.pullAndUpdateContainerReference(task, container)
+}
+
+func (engine *DockerTaskEngine) pullAndUpdateContainerReference(task *api.Task, container *api.Container) DockerContainerMetadata {
```

Comment:
Since there is already a write lock in updating(add/remove) the ContainerReference in imageState and a read lock for getting the imageState in imageManager, after going through the code, I don't think extra lock is needed here.